Cannot create full datetime product attribute 4053

novikor opened the pull request:
<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description (*)
  -  Added time picker for product edit page (for attrobutes with `frontend_type` = `date`)


### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/magento2#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->
1. magento/magento2#4053: Cannot create full datetime product attribute

### Manual testing scenarios (*)
1. Go to product edit page
2. Find any `date` attribute
3. Try to save date & time value
4. Make sure value is saved correctly
### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

magento-engcom-team commented:
Hi @novikor. Thank you for your contribution 
Here is some useful tips how you can test your changes using Magento test environment.
Add the comment under your pull request to deploy test or vanilla Magento instance:
- `@magento-engcom-team give me test instance` - deploy test instance based on PR changes
- `@magento-engcom-team give me 2.3-develop instance` - deploy vanilla Magento instance

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation

sivaschenko commented on app/code/Magento/Ui/Component/Form/Element/DataType/DateTime.php:
```diff
@@ -0,0 +1,132 @@
+<?php
```
New added classes should have strict mode enabled, also, can you please cover this class with unit test

novikor replied:
@sivaschenko strict types have been added, - it is my fault, just copied class and forgot to refactor it.
I agree that Unit tests are a good thing, but, in fact, there is nothing to cover.

sivaschenko commented on app/code/Magento/Ui/Component/Form/Element/DataType/DateTime.php:
```diff
@@ -0,0 +1,132 @@
+            }
+            return $dateObj;
+        } catch (\Exception $e) {
+            return null;
```
I think it's better to log the exception message

novikor replied:
@sivaschenko, Please recheck

sivaschenko reviewed (dismissed):
@novikor Thank you for the contribution! Please see the review notes.

novikor commented:
@sivaschenko , any updates?

sivaschenko commented:
@novikor thanks for your updates, I need a bit more time to process this pull request. One thing that already araised - these changes should be definitely covered by integration tests. Can you please provide the coverage?

novikor commented:
@sivaschenko , will be done

novikor commented:
@sivaschenko , please check. I found only one place when attribute input types are being covered by integrational tests and covered it in the same way.

Also, it seems that functional tests failed because of some troubles with travis/licenses.

VladimirZaets commented:
Hi @novikor. Thanks for collaboration. Please explain why you need separate component for the datetime input if it is only one property of configuration option for date component? As for me, it's redundant, each of components have a lot of configuration options and creating separate components for each of them as for me is overhead.

novikor commented:
@VladimirZaets , I thought it was a good idea to leave existing component the same to avoid any effect on existing usage of it in case of Magento upgrade on any project.

If it does not have any sense, I will remake it these week. Please let me know.

Thank you for review.

novikor commented:
@VladimirZaets, I suggest also to mark the existing date component as deprecated and remove it in 2.4. So, only datetime component will be used in future.

novikor commented:
Hi, @VladimirZaets . I did it.

novikor committed:
Cannot create full datetime product attribute 4053.
Enabled time picker for product edit page.

VladimirZaets commented:
Hi @novikor. Great, thank you!

magento-engcom-team commented:
Hi @VladimirZaets, thank you for the review.
[ENGCOM-4117](https://jira.corp.magento.com/browse/ENGCOM-4117) has been created to process this Pull Request

VladimirZaets commented:
Hi @novikor. 

Your fix provide the next problem (manual testing scenario):
Create product "Test"
Click Edit for "Test" product
Set date for Set Product as New From
 Date is saved in format mm/dd/yyyy without time
Edit file app/code/Magento/Ui/view/base/web/js/lib/knockout/bindings/datepicker.js
Set showsTime: true
Clear pub/static and flush cache
Create product "Test1"
Click Edit for "Test1" product
Set date and time for Set Product as New From
 Date is saved in format mm/dd/yyyy 12:00 AM (time always is shown as 12:00 AM )

novikor commented:
@VladimirZaets , thanks for the update.
I will check it in the nearest future.

Commit:
Cannot create full datetime product attribute 4053.

Commit:
Merge branch '2.3-develop' into Cannot-create-full-datetime-product-attribute-4053

novikor commented:
@VladimirZaets , fixed.
I can`t imagine how could it work before.

VladimirZaets commented:
@novikor thanks

magento-engcom-team commented:
Hi @VladimirZaets, thank you for the review.
[ENGCOM-4117](https://jira.corp.magento.com/browse/ENGCOM-4117) has been created to process this Pull Request

VasylShvorak commented:
Testing result: Inconsistency between time in admin and time in database.
**Manual testing scenario:**

- Install magento with timezone `--timezone=America/Chicago` 
- Go to Catalog->Products
- Click Add Product
-- Fill required fileds
-- Set Product as New From = Go Today
- Save Product
- Compare time in admin and database
-  Time on product edit page is shown according timezone, time in database is shown according UTC(in my case)
![#19385Admin](https://user-images.githubusercontent.com/26908019/54677039-c1853700-4b0a-11e9-83ac-db3699a34765.png)
![#19385DB](https://user-images.githubusercontent.com/26908019/54677187-0dd07700-4b0b-11e9-8818-0a3a630206a7.png)


@sivaschenko @VladimirZaets Is this correct behavior?

sivaschenko commented:
@VasylShvorak that's correct as long as the time in UI is displayed consistentently.

VasylShvorak commented:
Hi @novikor Time is displayed inconsistently. Time is saved,for example, in timezone=America/Chicago by Calendar widget and time is displayed in UTC on Products grid page.
 ![#19385ProdEdit](https://user-images.githubusercontent.com/26908019/54901197-65415f00-4ede-11e9-82f6-b5397b601786.png) 
 ![#19385ProdGrid](https://user-images.githubusercontent.com/26908019/54901201-670b2280-4ede-11e9-8a70-b34217e10d79.png)

novikor commented:
I am working on it.

Commit:
Cannot create full datetime product attribute 4053.
Set timezone for date columns of catalog grid;
Created an extension point for that purpose

novikor commented:
Now correct timezone is set for `date` columns of catalog grid.
An extension point has been provided for that purpose.

Please review.

sivaschenko commented on app/code/Magento/Catalog/Ui/Component/Column/DateConfigProvider.php:
```diff
@@ -0,0 +1,55 @@
+     *
+     * @return array
+     */
+    public function getConfig(string $dataType):array
```
What is the purpose of `$dataType` argument for the concrete provider?

novikor replied:
Actually, there is no purpose. However, they should implement the same interface.
What about changing it to `string $dataType = null`?

sivaschenko replied:
@novikor I don't think the pool and the provider should implement the same interface in this case (as pool's interface does not make sense for the concrete provider)

novikor replied:
@sivaschenko , please recheck my solution.

sivaschenko commented on app/code/Magento/Catalog/Ui/Component/ColumnFactory.php:
```diff
@@ -43,11 +45,21 @@ class ColumnFactory
+     */
+    public function __construct(
+        \Magento\Framework\View\Element\UiComponentFactory $componentFactory,
+        DataTypeConfigProviderInterface $configProvider
```
```suggestion
        DataTypeConfigProviderInterface $configProvider = null
```

sivaschenko reviewed (dismissed):
Hi @novikor thanks for the fix and the extension point. Please see my review comments

sivaschenko committed:
Update app/code/Magento/Catalog/Ui/Component/ColumnFactory.php

Co-Authored-By: novikor <novik.kor@gmail.com>

novikor committed:
Cannot create full datetime product attribute 4053.
Pool became an API now.
The pool does not implements DataTypeConfigProviderInterface

VasylShvorak commented:
@magento-engcom-team give me test instance

magento-engcom-team commented:
Hi @VasylShvorak. Thank you for your request. I'm working on Magento instance for you

magento-engcom-team commented:
Hi @VasylShvorak, here is your [new Magento instance](https://pr-19385.instances.magento-community.engineering).
Admin access: [https://pr-19385.instances.magento-community.engineering/admin](https://pr-19385.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`

VasylShvorak commented:
Hi @novikor during testing PR changes I got an error: `Class Magento\Catalog\Ui\Component\DataTypeConfigProviderInterface does not exist` when navigate to Catalog->Products.
![#19385Exc](https://user-images.githubusercontent.com/26908019/57523347-6531dd00-732d-11e9-9775-8c62e5083f89.png)

Could you take a look?

novikor commented:
@VasylShvorak , please recheck.
It was a silly mistake from my side.

magento-engcom-team commented:
Hi @sidolov, thank you for the review.
[ENGCOM-4117](https://jira.corp.magento.com/browse/ENGCOM-4117) has been created to process this Pull Request

VasylShvorak commented:
:heavy_check_mark: QA passed

p-bystritsky commented:
Hello, @novikor. Please fix the test failures listed below.

sidolov commented:
@novikor , I am closing this PR now due to inactivity.
Please reopen and update if you wish to continue.
Thank you for the collaboration!

novikor committed:
Merge remote-tracking branch 'upstream/2.3-develop' into Cannot-create-full-datetime-product-attribute-4053

novikor committed:
Cannot create full datetime product attribute 4053.

novikor commented:
**Notice**: please recheck https://github.com/magento/magento2/issues/23113

novikor committed:
Cannot create full datetime product attribute 4053.
Fixed static tests

engcom-Delta commented:
Hi @novikor I cannot reproduce issue [#23113](https://github.com/magento/magento2/issues/23113) on 2.3-develop.
English locale:
![#19385English](https://user-images.githubusercontent.com/51681379/64410756-646b4680-d094-11e9-94e1-d3dddb127246.png)

French locale:
![#19385French](https://user-images.githubusercontent.com/51681379/64410769-6a612780-d094-11e9-9c69-f6c93cbc2f26.png)

sidolov commented:
Hi @novikor , since mentioned issue not reproducing I'm closing this PR. Feel free to reopen if you find one more case. Thank you for your contribution!
